### PR TITLE
Add Example ACGAN cifar-10 version

### PR DESCRIPTION
**Pull request by King-Of-Knights**
Modified from mnist_acgan. Use some tricks to combat with overfitting and disconverage. Hope you can verify this work!

### CODE REVIEW

**Comment by King-Of-Knights** [comment]
Result is here
https://github.com/King-Of-Knights/Keras-ACGAN-CIFAR10

**Comment by ozabluda** [comment]
There is no diversity whatsoever  - all images from the same class seem identical. And pretty bad too, made worse by impossibility to hunt for a couple of good ones (since they are all the same).

**Comment by King-Of-Knights** [comment]
@ozabluda Hello! thanks for your review! I have try some new tricks to avoid mode collapse,  hope u can recheck the result:
(https://github.com/King-Of-Knights/Keras-ACGAN-CIFAR10)
if it is ok, I will reopen this PR
Thanks for your time!

**Comment by ozabluda** [comment]
the results are much better now, but are still unrecognizable by a human. What did you change in an existing Keras MNIST example? The only obvious to me thing is stochastic two-sided smoothing.

**Comment by King-Of-Knights** [comment]
@ozabluda Thanks for your review, you said it! the most important trick is : Label soomthing.
Your Generator Loss just go up and go up, if you just change mnistdataset to cifar-10dataset in mnist-acgan.py. Discriminator just kill Generator, since G has no idea how to cheat unstoppable D 
 (Discriminator Overfitting), actually it has really afflicted me for a long time. This trick prevent D become too overconfidence.
The second trick is K times train D vs 1 time train G, since I use a Guass Noise Layer in D, D should be trained more to guide G.  3 times train D vs 1 time train G or  4 times train D vs 1 time train G is both OK, But 1 times train D vs 1 time train G and  2 times train D vs 1 time train G provides not good result. I have no idea why is 3, but it just works.
The third trick is Minibatch Discrimination. if you don't use this trick in your D, the "Airplane" and "Frog" class are easy to trapped in mode collapse even you choose the Hyper parameters mentioned in original [paper](https://arxiv.org/abs/1610.09585). use Minibatch Discrimination, things get much better!
"but are still unrecognizable by a human", of course! the generated images is not clear as the orignal images, I don't read many papers, but in the paper I read, people are not prone to share the result about Cifar-10, they will mention, but rarely give the generated images. I believe GAN in Cifar-10 result is not as good  as Human face, ImageNet cats and so on. Since the statistics information in Cifar-10 is not as other  strong as dataset. Actually if you could stare at the generated image for one more second while keep thinking what class it belongs to, I think it won't be very difficult to figure out some picture in it. Btw, real images in cifar-10 are little hard to tell what it is ,since it is just a low-resolution images!
Again, thanks for ur time, have a nice day! :)

**Comment by King-Of-Knights** [comment]
@ozabluda Any advise to improve ACGAN performance on cifar-10? Mnist is not a very hard job for GAN. But Cifar-10 is really a big challenge for GAN . If ACGAN works on Cifar-10, it should also works well on ImageNet. 👍

**Comment by ozabluda** [comment]
>Generator Loss just go up and go up, if you just change mnist dataset to cifar-10 dataset in mnist-acgan.py. Discriminator just kill Generator, since G has no idea how to cheat unstoppable D

Does it still happen with current master version? It was a problem even on MNIST, but it was fixed in PR https://github.com/keras-team/keras/pull/8383.

If yes, what about with not yet merged https://github.com/keras-team/keras/pull/8616?

**Comment by King-Of-Knights** [comment]
@ozabluda I will try and report back!

**Comment by King-Of-Knights** [comment]
@ozabluda I try your advice, but no luck, just got the black image, looks like it doesn't converge.
I just update the result ! That's the best result so far.
I think that is almost the result that can't be surpass. In paper, the ImageNet doesn't seem good in animal. It looks like GAN has extracted the correct feature, but couldn't construct it correctly, so does this happend to cifar-10 cat, dog  and deer . I will upload a zip contains 1-150th epoch generate images [here](https://drive.google.com/open?id=1ssCCu_T8VIgF4B2lfF5yRixB9MCbn1NL), so you can take a quick overview it
 Somehow I disagree #8482 , in paper formula(2)(3) aux classifier should be trained on generated image

**Comment by ozabluda** [comment]
Sorry, I don't understand. https://drive.google.com/open?id=1ssCCu_T8VIgF4B2lfF5yRixB9MCbn1NL are not all black. Or these are the updated results? Updated how?

>Somehow I disagree #8482 , in paper formula(2)(3) aux classifier should be trained on generated image also I observe an interesting phenomena, Image quality will improve, if the aux classifier accuracy decrease a little, just like what happen in Improve gan

See second paragraph of https://github.com/keras-team/keras/pull/8482#issuecomment-345766120 and links there, for discussion of why it was done. Also see https://github.com/keras-team/keras/pull/8452#issuecomment-343705650

**Comment by King-Of-Knights** [comment]
@ozabluda hello! update means I update some hype parameter in my original code got better picture.  Got black image in the master version  thanks for review, I have no idea how to improve this code now. maybe I should reopen this to let more people see this. maybe a better idea will come out

**Comment by ozabluda** [comment]
sure

**Comment by King-Of-Knights** [comment]
@ozabluda I will try one by one.  Actually,  some people just set the learning rate from 0.002 to 0.00002 in mnist_acgan.py fix the black image. but it's not working for cifar-10 version

**Comment by fchollet** [comment]
@ozabluda could you please review this PR? Will merge when you LGTM.

**Comment by ozabluda** [comment]
@fchollet, in progress, see all my previous comments in this PR.

**Comment by fchollet** [comment]
Is this PR abandoned?

**Comment by fchollet** [comment]
Closing seemingly abandoned PR. Feel free to reopen with an update.
